src/qforte/utils/symmetry_analysis.py

```diff
@@ -0,0 +1,131 @@
+    N.add_term(n_qubits / 2, circ_const)
+    for qubit in range(n_qubits):
+        circ_Z = qf.Circuit()
+        circ_Z.add_gate(qf.gate('Z', qubit))
+        N.add_term(-0.5, circ_Z)
+    N_sqrd = qf.QubitOperator()
+    N_sqrd.add_op(N)
+    N_sqrd.operator_product(N, True, True)
+    N_exp_val = qc.direct_op_exp_val(N)
+    N_var = qc.direct_op_exp_val(N_sqrd) - N_exp_val**2
+    if np.imag(N_exp_val) != 0 or np.imag(N_var) != 0:
+        raise ValueError('The expectation value and variance of the particle number should be real!')
+    N_exp_val = np.real(N_exp_val)
+    N_var = np.real(N_var)
+
+    # Compute <Sz> and <Sz^2> - <Sz>^2, where Sz is the projection of the total spin on the z axis
+    Sz = qf.total_spin_z(n_qubits)
+    Sz_sqrd = qf.QubitOperator()
+    Sz_sqrd.add_op(Sz)
+    Sz_sqrd.operator_product(Sz, True, True)
+    Sz_exp_val = qc.direct_op_exp_val(Sz)
+    Sz_var = qc.direct_op_exp_val(Sz_sqrd) - Sz_exp_val**2
+    if np.imag(Sz_exp_val) != 0 or np.imag(Sz_var) != 0:
+        raise ValueError('The expectation value and variance of Sz should be real!')
+    Sz_exp_val = np.real(Sz_exp_val)
+    Sz_var = np.real(Sz_var)
+
+    # Compute <S^2> and <S^4> - <S^2>^2, where S is the total spin
+    S_sqrd = qf.total_spin_squared(n_qubits)
+    S_sqrd_sqrd = qf.QubitOperator()
+    S_sqrd_sqrd.add_op(S_sqrd)
+    S_sqrd_sqrd.operator_product(S_sqrd, True, True)
+    S_sqrd_exp_val = qc.direct_op_exp_val(S_sqrd)
+    S_sqrd_var = qc.direct_op_exp_val(S_sqrd_sqrd) - S_sqrd_exp_val**2
+    if np.imag(S_sqrd_exp_val) !=0 or np.imag(S_sqrd_var) !=0:
+        raise ValueError('The expectation value and variance of S^2 should be real!')
+    S_sqrd_exp_val = np.real(S_sqrd_exp_val)
+    S_sqrd_var = np.real(S_sqrd_var)
+
+    # Compute the contributions of various symmetry spaces to the wavefunction
+    weight_particle_number = [0] * (n_qubits + 1)
+    weight_Sz = [0] * (n_qubits + 1)
+    Sz_values = []
+    Sz_values_dictionary = {}
+    for i in range(n_qubits + 1):
+        Sz_values.append(-n_qubits * 0.25 + 0.5 * i)
+        Sz_values_dictionary[Sz_values[i]] = i
+    weight_irreps = [0] * len(irreps)
+    for det in range(1<<n_qubits):
+        occ_alpha = []
+        occ_beta = []
+        for i in range(0, n_qubits, 2):
+            if (1<<i)&det != 0:
+                occ_alpha.append(i)
+            if (1<<(i+1))&det != 0:
+                occ_beta.append(i+1)
+        weight = qc.get_coeff_vec()[det]
+        weight = weight * np.conjugate(weight)
+        particle_number = len(occ_alpha + occ_beta)
+        Sz_value = (len(occ_alpha) - len(occ_beta)) * 0.5
+        irrep = qf.sq_op_find_symmetry(orb_irreps_to_int, occ_alpha + occ_beta, [])
+        weight_particle_number[particle_number] += weight
+        weight_Sz[Sz_values_dictionary[Sz_value]] += weight
+        weight_irreps[irrep] += weight
+    for i in range(n_qubits + 1):
+        weight_particle_number[i] = np.real(weight_particle_number[i])
```

Comment:
l. 88 should guarantee these are real, so why is this necessary? At _most_, just call `np.real` on l. 88.

Reply:
This is done purely for printing purposes: I want to avoid printing 0.0000000000j all the time. I will follow your suggestion and call `np.real` on line 88.